src/order.c

```diff
@@ -0,0 +1,4148 @@
+
+/*
+ * `int_order_impl()` is used by both `int_order()` and by
+ * `chr_order()`.
+ *
+ * For `chr_order()`, the TRUELENGTH values will already be extracted into
+ * `p_lazy_x_chunk`. In this case, we set `copy = false` to tell
+ * `int_order_impl()` to use `p_lazy_x_chunk` directly rather than copying `p_x`
+ * over to `p_x_chunk`.
+ */
+static
+void int_order_impl(const int* p_x,
+                    struct lazy_order* p_lazy_o,
+                    struct lazy_vec* p_lazy_x_chunk,
+                    struct lazy_vec* p_lazy_x_aux,
+                    struct lazy_vec* p_lazy_o_aux,
+                    struct lazy_vec* p_lazy_bytes,
+                    struct lazy_vec* p_lazy_counts,
+                    struct group_infos* p_group_infos,
+                    bool decreasing,
+                    bool na_last,
+                    R_xlen_t size,
+                    bool copy) {
+  if (size <= INSERTION_ORDER_BOUNDARY) {
+    lazy_order_initialize(p_lazy_o);
+    int* p_o = p_lazy_o->p_o;
+
+    void* p_x_chunk = int_maybe_copy(p_x, p_lazy_x_chunk, size, copy);
+    int_adjust(p_x_chunk, decreasing, na_last, size);
+
+    int_insertion_order(p_x_chunk, p_o, p_group_infos, size);
+
+    return;
+  }
+
+  uint32_t range;
+  int x_min;
+
+  int_compute_range(p_x, size, &x_min, &range);
+
+  /*
+   * If in counting order range and on the first column / single vector,
+   * `p_o_aux` won't be used, so no need to initialize it.
+   *
+   * Also, `p_o` will be filled directly, so for performance we don't
+   * initialize its order.
+   */
+  if (range < INT_COUNTING_ORDER_RANGE_BOUNDARY) {
+    const bool initialized = false;
+
+    int* p_o = p_lazy_o->p_o;
+    int* p_o_aux = (int*) p_lazy_o_aux->p_data;
+
+    int_counting_order(
+      p_x,
+      p_o,
+      p_o_aux,
+      p_group_infos,
+      size,
+      x_min,
+      range,
+      initialized,
+      decreasing,
+      na_last
+    );
+
+    p_lazy_o->initialized = true;
+    return;
+  }
+
+  lazy_order_initialize(p_lazy_o);
+  int* p_o = p_lazy_o->p_o;
+
+  lazy_vec_initialize(p_lazy_o_aux);
+  int* p_o_aux = (int*) p_lazy_o_aux->p_data;
+
+  lazy_vec_initialize(p_lazy_x_aux);
+  uint32_t* p_x_aux = (uint32_t*) p_lazy_x_aux->p_data;
+
+  lazy_vec_initialize(p_lazy_bytes);
+  uint8_t* p_bytes = (uint8_t*) p_lazy_bytes->p_data;
+
+  lazy_vec_initialize(p_lazy_counts);
+  R_xlen_t* p_counts = (R_xlen_t*) p_lazy_counts->p_data;
+  memset(p_counts, 0, p_lazy_counts->size);
+
+  void* p_x_chunk = int_maybe_copy(p_x, p_lazy_x_chunk, size, copy);
+  int_adjust(p_x_chunk, decreasing, na_last, size);
+
+  int_radix_order(
+    p_x_chunk,
+    p_o,
+    p_x_aux,
+    p_o_aux,
+    p_bytes,
+    p_counts,
+    p_group_infos,
+    size
+  );
+}
+
+// If we aren't copying, we expect that `p_lazy_x_chunk` already contains
+// the integer values. This happens for character ordering where we store
+// the truelength values in `p_lazy_x_chunk`.
+static inline
+void* int_maybe_copy(const int* p_x,
```

Comment:
Hmm I don't think storing the `->type_size` would work. `p_x_chunk` is reused between columns, but it might be used for a double column at one point, then for an integer column later on. It is allocated to be the size of the largest column, but storing that information isn't necessarily useful because it might be used on a smaller column type.

Reply:
That makes sense.

I think the chunk pointer should be an `unsigned char*` rather than a `void*`. The latter makes it seem like it has one single effective type, which isn't the case.